src/Components/Components/src/ComponentBase.cs

```diff
@@ -177,188 +177,99 @@ void IComponent.Configure(RenderHandle renderHandle)
-            if (!_hasCalledInit)
+            if (!_initialized)
             {
-                return RunInitAndSetParameters();
+                _initialized = true;
+
+                return RunInitAndSetParametersAsync();
             }
             else
             {
-                OnParametersSet();
-                // If you override OnInitAsync or OnParametersSetAsync and return a noncompleted task,
-                // then by default we automatically re-render once each of those tasks completes.
-                var isAsync = false;
-                Task parametersTask = null;
-                (isAsync, parametersTask) = ProcessLifeCycletask(OnParametersSetAsync());
-                StateHasChanged();
-                // We call StateHasChanged here so that we render after OnParametersSet and after the
-                // synchronous part of OnParametersSetAsync has run, and in case there is async work
-                // we trigger another render.
-                if (isAsync)
-                {
-                    return parametersTask;
-                }
-
-                return Task.CompletedTask;
+                return CallOnParametersSetAsync();
             }
         }
 
-        private async Task RunInitAndSetParameters()
+        private async Task RunInitAndSetParametersAsync()
         {
-            _hasCalledInit = true;
-            var initIsAsync = false;
-
             OnInit();
-            Task initTask = null;
-            (initIsAsync, initTask) = ProcessLifeCycletask(OnInitAsync());
-            if (initIsAsync)
+            var task = OnInitAsync();
+
+            if (task.Status != TaskStatus.RanToCompletion && task.Status != TaskStatus.Canceled)
             {
                 // Call state has changed here so that we render after the sync part of OnInitAsync has run
                 // and wait for it to finish before we continue. If no async work has been done yet, we want
                 // to defer calling StateHasChanged up until the first bit of async code happens or until
-                // the end.
+                // the end. Additionally, we want to avoid calling StateHasChanged if no
+                // async work is to be performed.
                 StateHasChanged();
-                await initTask;
-            }
 
-            OnParametersSet();
-            Task parametersTask = null;
-            var setParametersIsAsync = false;
-            (setParametersIsAsync, parametersTask) = ProcessLifeCycletask(OnParametersSetAsync());
-            // We always call StateHasChanged here as we want to trigger a rerender after OnParametersSet and
-            // the synchronous part of OnParametersSetAsync has run, triggering another re-render in case there
-            // is additional async work.
-            StateHasChanged();
-            if (setParametersIsAsync)
-            {
-                await parametersTask;
+                await ProcessLifecycleTaskAsync(task);
```

Comment:
This introduces an extra StateHasChanged event. We want StateHasChanged to be called right before every "true" async operation and after the sum of the sync operations and the "sync" part of an async operation runs.

For example:
OnInit
OnInitAsync (Completely synchronous)
OnParametersSet()
OnParametersSetAsync (Completely synchronous)
StateHasChanged()

OnInit
OnInitAsync (Completely synchronous)
OnParametersSet()
OnParametersSetAsync (Sync)
StateHasChanged()
OnParametersSetAsync (Continuation)
StateHasChanged()